### PR TITLE
Initial support/tests for SQLite flags and db dump

### PR DESCRIPTION
**Pull request by benjolitz**
This commit adds a pass through for specifying SQLite3 flags. This is
an absolute requirement for being able to create in-memory DB
representations.

### CODE REVIEW

**Comment by benjolitz** [comment]
Travis is busted, not the change set:
```
The following NEW packages will be installed:
  crystal
0 upgraded, 1 newly installed, 0 to remove and 241 not upgraded.
Need to get 11.9 MB of archives.
After this operation, 36.4 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

**Comment by asterite** [comment]
@benjolitz Hi. Thanks!

I don't know much about sqlite3, but you can already use in-memory databases. For example this works:

```crystal
require "sqlite3"

SQLite3::Database.new(":memory:") do |db|
  db.execute "create table person (name string, age integer)"
  db.execute %(insert into person values ("foo", 10))
  db.query("select * from person") do |result_set|
    puts result_set.next
    puts result_set["name"]
    puts result_set["age"]
  end
end
```

So why are the flags needed, and what do they do?

In any case, if we need a second constructor with flags, I'd move the flags into an enum outside lib: lib code (or enums, for that matter) shouldn't be needed in daily use. That enum would only include the values that make sense for open (I see there are other values that are used for other things).

**Comment by benjolitz** [comment]
I have a 8 gigabyte SQLite database that yields query sets of between 10-100mb at a time. Those result sets take 8 seconds per request to completion.

If you do the above suggestion, every query will block on the prior one. Thus for three concurrent requests, they will return in 8, 16, 24 seconds.

(This is while using channels to send an encoded row to the HTTP response function instance)

As there is one table and you cannot query the same table while another query is running, a single memory DB is patently useless.

Cloning the DB is also useless, as every clone would take another 8 gigabytes.

The official solution, as suggested by https://www.sqlite.org/inmemorydb.html , is to create a read-write memory database and share it's cache and access to multiple read-only memory databases. They will not block.

The flags allow me to:

A) Create a primary memory database in a shared cache mode using SQLite3's URI (https://www.sqlite.org/sharedcache.html)

B) Create a READONLY memory DB slave that uses the primary's cache which allows for parallel queries (these are bound to a request)

C) Put all DB's in FULL_MUTEX mode which allows for multiple concurrent reads

With the above, the requests all complete at 24 seconds, allowing the scheduler and a Fiber.current.yield to give fairness/quality of service across the requests.

":memory:" disables shared cache and simply is not good enough.

I highly disagree with the interest in restricting SQLite3 flags. Most use cases won't touch those flags. The ones that do, however, often know what they're doing.

If I were to suggest limiting the API, I would put forth a dedicated SQLite3 URI support, but that requires that the driver now know how to match flags to URI fragments.

Finally, it is not possible as-is to put the database either into read-only mode nor enable mutex locking. Any "helpful" interface still would need to expose toggleables.

By creating a simplified pass through to controlling `sqlite3_open_v2` flags, this places the responsibility onto the developer who is asking for special case behavior to make it work.

**Comment by asterite** [comment]
Thank you for the wonderful explanation!!

Let's add this functionality then.

What I meant with the enum is that I don't want users of this API to know about `LibSQLite3`, because it's not documented, and libs are usually used internally and a wrapper API uses them.

So, we can do this. Right now the enum is defined in the lib, [here](https://github.com/manastech/crystal-sqlite3/blob/master/src/sqlite3/lib_sqlite3.cr#L8). We can create a separate `enums.cr` file and define it there, inside the `SQLite3` module:

```crystal
module SQLite3
  enum Flag
    READONLY = 0x00000001
    # etc.
  end
end
```

then do `require "./enums"` in the `lib` and use that type. Then in the Database API use `SQLite3::Flag` instead of `LibSQLite3::Flag`.

The second thing is that I see some [comments](LibSQLite3) in the Flag enum, like "Ok for sqlite3_open_v2()" and "VFS only". I don't know about this either, but is it maybe possible that the "VFS only" don't make sense for create? That's why I wanted to remove those values from the flags passed to create. However, this is just a minor detail, I wouldn't mind (at all!) just exposing all values, because the user using those flags probably knows what he's doing (like in your case :smile:).

In conclusion: what do you think above moving the enum from the lib to the module?

**Comment by benjolitz** [comment]
"VFS only" options mean that they only apply to a real file (i.e. not a ":memory:" database).

"Ok for sqlite3_open_v2" means that the option, which applies for `sqlite3_open` also does what is expected with `sqlite3_open_v2` (which is what this driver uses)

I like your idea of a `SQLite3::Flag` setup -- it would make my code less "noisy" (i.e. `SQLite3::LibSQLite3::Flag`), so I implemented it.

Regarding VFS/"Okay for sqlite3_open_v2" -- there's two suggestions:

1) Let the developer decide their options. If it blows up, it's their fault.

2) We separate out the VFS options into a `VFSFlags` enum.
2.1) We then check if any VFSFlags are inside a group of `Flags`
2.2) If the above is true, then we check if the filename/url_path is /not/ a memory db.

1 is easier for now (as an early product/language, it's Okay for Crystal to behave like that)

2 is what I'd do if this driver became super mainstream and was used by students (who need all the help they can get)

I suggest 1 due to time constraints. But I'll think about 2 -- it requires one to fully support arbitrary SQLite3 URIs and compose the correct flags as required by the URI.

**Comment by benjolitz** [comment]
I would like to note that the Travis build is still broken.

Someone needs it to do `sudo apt-get -y install crystal` (the -y option would avoid the infinite prompt hang for "Y/N?")

**Comment by asterite** [comment]
Thanks, now the changes are perfect! We can discuss 2 in another issue later, but for now I think this is very good and probably 2 won't be needed :-)

**Comment by asterite** [comment]
I'll try to see what's wrong with travis...
